STONEBLD-75: Populate build-service.md (#67)
redhat-appstudio/book: book/build-service.md

# Build Service

## Overview

The Build Service is composed of controllers that create and configure build pipelines. The main input for the Build Service is a Component CR managed by the [Hybrid Application Service (HAS)](hybrid-application-service.md).

> ![](../diagrams/build-service/build-service.drawio.svg)

Reviewer: I think the green and red arrows in the diagram may be reversed. Green is default and red is custom?

Author: It's written in diagram at bottom

Author: ok, I cannot read, fixing ;-)

### Dependencies

The Build Service is dependent on the following services:
- [Pipeline Service](./pipeline-service.md)
  - Pipeline execution, Pipeline logging
- [Hybrid Application Service](./hybrid-application-service.md)
  - Provides Component CR with annotations and `.status.devfile` which is used for PipelineRun configuration.
- [Image Controller](./image-controller.md)
  - Generation of container image repository which is used by PipelineRun

## Controllers

> The Build Service contains these controllers:

Reviewer 2: Where does the image controller land? Is that within the build service or outside it?

Reviewer 2: Nevermind, I see that it is another top-level document (currently empty) so I assume that it falls outside of the build service.

Author: Added image controller into diagram

Reviewer 2: Thanks for the image. Is there any intention to populate `image-controller.md` in this PR?

Author: Absolutely not.

Reviewer 2: That's fine. Can you put some "content" into it as a TODO/nod to future work/intentionally left blank?

Author: I used same approach as it was done before:
https://github.com/redhat-appstudio/book/blob/main/book/build-service.md
https://github.com/redhat-appstudio/book/blob/main/book/hybrid-application-service.md
https://github.com/redhat-appstudio/book/blob/main/book/gitops-service.md
https://github.com/redhat-appstudio/book/blob/main/book/java-rebuilds-service.md
https://github.com/redhat-appstudio/book/blob/main/book/service-provider-integration.md
https://github.com/redhat-appstudio/book/blob/main/book/workspace-and-terminal-service.md

- Component Build Controller

>   - Used for PipelineRuns creation.

Reviewer 3: Does it make sense to extend this?

- PaC PipelineRun Pruner Controller
  - Deletes PipelineRuns managed by Pipelines As Code for Components that were deleted.
- Git Tekton Resources Renovater
  - Provides updates for the .tekton directory in the user repository which has been created by Component Build Controller in Custom Mode.

### Component Build Controller

Component Build Controller is managed by Component CR changes. It's using Component CR annotations and `.status.devfile` for selecting a mode and configuration of the PipelineRuns.

#### Modes

Component Build Controller is working in two modes:
- Default Mode

>   - Mode where PipelineRuns are created directly by the Build Service.

Reviewer 3: I'd emphasise that the pipeline run is created directly by the controller.

Author: made bold `directly`

>   - Simple mode where PipelineRuns are created directly by the Build Service.

Reviewer: In a few places, this is called "simple mode" and in a few other places this is called "default mode".

Should we consolidate to using just one term or the other?

Reviewer 2: fwiw, I prefer default mode, but I don't know if there is a precedence for simple elsewhere.

Author: sure, updated to Default mode to match with HAC

- Custom Mode

>   - Integrates with Pipelines-As-Code, the Component Build Controller creates PipelineRuns definitions in the user code repository.

Reviewer 4: Probably we should use a unified way to describe "Pipelines as Code" and avoid using PaC in this document.

Pipelines as Code website uses Pipelines-as-Code, and from my point of view, "Pipelines as Code" is ok as well. Whichever is selected, just use it consistently.

>   - The prerequisite is to have installed GitHub App which is used by the Build Server in the user's repository.

Reviewer 4: What is Build Server?

Should it `s/Server/Service/`?

Default mode:
1. Wait until the component has `.status.devfile` set
1. Get data from `.status.devfile` and use them for creating PipelineRun
1. Set annotation on the component - `appstudio.openshift.io/component-initial-build: processed`

Delete annotation `appstudio.openshift.io/component-initial-build` on Component to retrigger PipelineRun in Default mode.

For the selection of Custom mode the Component CR must have set annotation `appstudio.openshift.io/pac-provision` to `request`.

Custom Mode:
1. Wait until the component has `.status.devfile` set
1. Get data from `.status.devfile`, use them for creating `.tekton` folder in the user's repository and create Repository CR so that Pipelines-as-Code is configured to monitor the user's repository.
1. Change the value of `appstudio.openshift.io/pac-provision` annotation to `done` in case the previous step was successful, in case of an issue the annotation is set to `error` and annotation `appstudio.openshift.io/pac-provision-error` with more details about the error is added.

#### PipelineRun selection

> The Build Service owns BuildPipelineSelector CRD which defines which PipelineRun to select for the Component CR. By default global BuildPipelineSelector `build-pipeline-selector` in `build-service` namespace is used. BuildPipelineSelector CR contains selectors, the first matching selector is used for the Component CR. The list of selectors can be extended by creating BuildPipelineSelector CR named the same as Application CR in the user's namespace so that will be applied to Components CRs under the Application CR. The list of selectors can be also extended for the whole user namespace by creating BuildPipelineSelector CR named `build-pipeline-selector` in the user's namespace.

Reviewer: Add a reference to https://redhat-appstudio.github.io/book/ref/build-service.html#buildpipelineselector somewhere here.

Author: added

> The Build Service owns [BuildPipelineSelector CRD](https://redhat-appstudio.github.io/book/ref/build-service.html#buildpipelineselector) which defines which PipelineRun to select for the Component CR. By default global BuildPipelineSelector `build-pipeline-selector` in `build-service` namespace is used. BuildPipelineSelector CR contains selectors, the first matching selector is used for the Component CR. The list of selectors can be extended by creating BuildPipelineSelector CR named the same as Application CR in the user's namespace so that will be applied to Components CRs under the Application CR. The list of selectors can be also extended for the whole user namespace by creating BuildPipelineSelector CR named `build-pipeline-selector` in the user's namespace.

Reviewer 5: `The list of selectors can be extended by creating BuildPipelineSelector CR named the same as Application CR in the user's namespace so that will be applied to Components CRs under the Application CR.`
seems too long, personally I would split it into two sentences:
`The list of selectors can be extended by creating BuildPipelineSelector CR with the same name as the Application CR in the user's namespace. This will ensure that it is applied to Component CRs under the corresponding Application CR.`

> The Build Service owns [BuildPipelineSelector CRD](https://redhat-appstudio.github.io/book/ref/build-service.html#buildpipelineselector) which defines which PipelineRun to select for the Component CR. By default global BuildPipelineSelector `build-pipeline-selector` in `build-service` namespace is used. BuildPipelineSelector CR contains selectors, the first matching selector is used for the Component CR. The list of selectors can be extended by creating BuildPipelineSelector CR named the same as Application CR in the user's namespace so that will be applied to Components CRs under the Application CR. The list of selectors can be also extended for the whole user namespace by creating BuildPipelineSelector CR named `build-pipeline-selector` in the user's namespace.

Reviewer 5: ```suggestion
The Build Service owns [BuildPipelineSelector CRD](https://redhat-appstudio.github.io/book/ref/build-service.html#buildpipelineselector) which defines which PipelineRun to select for the Component CR. By default global BuildPipelineSelector `build-pipeline-selector` in `build-service` namespace is used. BuildPipelineSelector CR contains selectors, the first matching selector is used for the Component CR. The list of selectors can be extended by creating BuildPipelineSelector CR with the same name as the Application CR in the user's namespace. This will ensure that it is applied to Component CRs under the corresponding Application CR. The list of selectors can be also extended for the whole user namespace by creating BuildPipelineSelector CR named `build-pipeline-selector` in the user's namespace.
```

> The Build Service owns [BuildPipelineSelector CRD](https://redhat-appstudio.github.io/book/ref/build-service.html#buildpipelineselector) which defines which PipelineRun to select for the Component CR. By default global BuildPipelineSelector `build-pipeline-selector` in `build-service` namespace is used. BuildPipelineSelector CR contains selectors, the first matching selector is used for the Component CR. The list of selectors can be extended by creating BuildPipelineSelector CR with the same name as the Application CR in the user's namespace. This will ensure that it is applied to Component CRs under the corresponding Application CR. The list of selectors can be also extended for the whole user namespace by creating BuildPipelineSelector CR named `build-pipeline-selector` in the user's namespace.

Reviewer 3: Would be helpful to mention that BuildPipelineSelector CR with the same name as the Application CR has higher priority than BuildPipelineSelector CR named `build-pipeline-selector` in the user's namespace, but they are not mutually exclusive and could be used together.

Author: Added in which order it's processed.

Example of BuildPipelineSelector:
```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: BuildPipelineSelector
metadata:
  name: build-pipeline-selector
  namespace: my-tenant
spec:
  selectors:
  - name: my-component
    pipelineRef:
      bundle: quay.io/my-user/my-bundle:v1.0
      name: my-bundle
    pipelineParams:
    - name: test
      value: fbc
    when:
      language: java
      componentName: my-component
      projectType: springboot
  - name: Docker build
    pipelineRef:
      bundle: quay.io/redhat-appstudio-tekton-catalog/pipeline-docker-build:3649b8ca452e7f97e016310fccdfb815e4c0aa7e
      name: docker-build
    when:
      dockerfile: true
  - name: NodeJS
    pipelineRef:
      bundle: quay.io/redhat-appstudio-tekton-catalog/pipeline-nodejs-builder:3649b8ca452e7f97e016310fccdfb815e4c0aa7e
      name: nodejs-builder
    when:
      language: nodejs,node

> ```

Reviewer 3: Please add a fallback case without `when` statement.
Also, an example of matching label (maybe with OR condition for values) would be useful as well.

Author: added

- output-image - generated by the Build Service the image repository is taken from Component CR's.
  - for Default mode, the tag is `initial-build-$TIMESTAMP-$RANDOM_NUMBER"`
  - for Custom mode, the tag is `{{ revision }}` which is evaluated by Pipelines As Code to commit SHA.
- docker-file - path to Dockerfile, only used when dockerfile is detected in Component's devfile
- path-context - context path for docker build, only used when dockerfile is detected in Component's devfile
- skip-checks - set to 'true' when Component contains annotation `skip-initial-checks: true`

### PaC PipelineRun Pruner Controller

In Default mode the PipelineRuns are connected to Component CR using OwnerRef, which handles the deletion of the PipelineRun CR when Component CR is deleted.

In Custom mode the PipelineRuns creation is handled by Pipelines as Code and there is no cleanup of the PipelineRuns when the Repository CR which manages the git repository is deleted.

The purpose of the PaC PipelineRun Pruner Controller is to remove the PipelineRun CRs created for Component CR which is being deleted.

### Git Tekton Resources Renovater

> In Custom mode the PipelineRun definition is created in the user's git repository under `.tekton` folder, this file is generated by Component Build Controller during Component CR creation. Git Tekton Resources Renovater is taking care of updating content `.tekton` during the lifetime of the Component CR.

Reviewer 4: s/content `.tekton`/content of `.tekton`/   ?

It would be helpful to mention that what content is updated by Renovator currently.

Example of BuildPipelineSelector:
```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: BuildPipelineSelector
metadata:
  name: build-pipeline-selector
  namespace: my-tenant
spec:
  selectors:
  - name: my-component
    pipelineRef:
      bundle: quay.io/my-user/my-bundle:v1.0
      name: my-bundle
    pipelineParams:
    - name: test
      value: fbc
    when:
      language: java
      componentName: my-component
      projectType: springboot
  - name: Label section
    pipelineRef:
      bundle: quay.io/my-user/my-bundle:v1.0
      name: labelled
    when:
      labels:
        mylabel: test
  - name: Docker build
    pipelineRef:
      bundle: quay.io/redhat-appstudio-tekton-catalog/pipeline-docker-build:3649b8ca452e7f97e016310fccdfb815e4c0aa7e
      name: docker-build
    when:
      dockerfile: true
  - name: NodeJS
    pipelineRef:
      bundle: quay.io/redhat-appstudio-tekton-catalog/pipeline-nodejs-builder:3649b8ca452e7f97e016310fccdfb815e4c0aa7e
      name: nodejs-builder
    when:
      language: nodejs,node
  - name: Fallback
    pipelineRef:
      bundle: quay.io/my-user/my-fallback:v1.0
      name: fallback
```

> In the example the first selector will match only when the component CR name is `my-component` and it's defined in the devfile or detected with `language: java` and `projectType: springboot` if all requests are matching then PipelineRun is generated based on `pipelineRef` and pipelineRun parameters are set to pipelineParams. If any of `when` condition does not match then it's checking the next selector. The `Fallback` sector does not contain `when` so it will be processed in case previous selectors do not match.

Reviewer 6: I am wondering .. is there a `Fallback` sector for the global BuildPipelineSelector CR`build-pipeline-selector` in `build-service` namespace in case there is not a match, how the scenario is handled and will there be any error visible to the user or suggested to define BuildPipelineSelector in the user's namespace?

Author: Right now it will fallback to hardcoded pipeline in build service. I added line with this informatoion.

Reviewer 6: Thank you!

- componentName
  - Name of Component CR
- projectType
  - projectType from devfile .metadata.projectType, can contain multiple values separated by a comma.
- dockerfile
  - if the Docker file is defined in generated devfile. Boolean value.
- annotations
  - match by Component CR annotations, expects map.
- labels
  - match by Component CR labels, expects map.

There are a few parameters that are set in PipelineRun created by the Build Service:
- git-url - taken from Component CR's spec.source.gitSource.url
- revision - taken from Component CR's spec.source.gitSource.revision when it's defined, otherwise it's using the default branch of the git repository
- output-image - generated by the Build Service the image repository is taken from Component CR's.
  - for Default mode, the tag is `initial-build-$TIMESTAMP-$RANDOM_NUMBER"`
  - for Custom mode, the tag is `{{ revision }}` which is evaluated by Pipelines As Code to commit SHA.
- docker-file - path to Dockerfile, only used when dockerfile is detected in Component's devfile
- path-context - context path for docker build, only used when dockerfile is detected in Component's devfile

> - skip-checks - set to 'true' when Component contains annotation `skip-initial-checks: true`

Author: added

### PaC PipelineRun Pruner Controller

In Default mode the PipelineRuns are connected to Component CR using OwnerRef, which handles the deletion of the PipelineRun CR when Component CR is deleted.

In Custom mode the PipelineRuns creation is handled by Pipelines as Code and there is no cleanup of the PipelineRuns when the Repository CR which manages the git repository is deleted.

The purpose of the PaC PipelineRun Pruner Controller is to remove the PipelineRun CRs created for Component CR which is being deleted.

### Git Tekton Resources Renovater

In Custom mode the PipelineRun definition is created in the user's git repository under `.tekton` folder, this file is generated by Component Build Controller during Component CR creation. Git Tekton Resources Renovater is taking care of updating content of `.tekton` folder during the lifetime of the Component CR.

Git Tekton Resources Renovater triggers the creation of Job CRs when BuildPipelineSelector `build-pipeline-selector` in `build-service` namespace is updated.

Workflow:

> 1. Get repositories installed for the GitHub Application matching to Components CRs and get tokens to access them.

Reviewer 6: One more question .. Does the `renovate` GH App needs to be installed in the user's github repository in order to use this cleanup feature?

Author: No we are using global application which is used by Build-service and pipelines-as-code.

Reviewer 6: Okay got it
